ichnaea/data/tests/test_stats.py

```diff
@@ -167,3 +167,10 @@ def test_update(self, celery, session):
                 assert values == (0, 0, 0, 3, 0)
             elif stat.region == "US":
                 assert values == (0, 0, 0, 0, 6)
+
+    def test_update_no_change(self, celery, session):
+        """update_statregion is successful when there are no area changes."""
+        self.test_update(celery, session)
+        update_statregion.delay().get()
+        stats = session.query(RegionStat).all()
+        assert len(stats) == 4
```

Comment:
Instead of calling `test_update` which then ties this test to that one, can we just do a minimal version of `test_update` in this test? Maybe something like this?:

```
   def test_update_no_change(self, celery, session):
        """update_statregion is successful when there are no area changes."""
        area = CellAreaFactory(radio=Radio.gsm, num_cells=1)
        area.region = None
        BlueShardFactory.create_batch(2, region="CA")
        BlueShardFactory.create_batch(3, region="GB")
        CellAreaFactory(radio=Radio.gsm, region="DE", num_cells=1)
        CellAreaFactory(radio=Radio.gsm, region="DE", num_cells=2)
        CellAreaFactory(radio=Radio.gsm, region="CA", num_cells=2)
        CellAreaFactory(radio=Radio.wcdma, region="DE", num_cells=3)
        CellAreaFactory(radio=Radio.lte, region="CA", num_cells=4)
        WifiShardFactory.create_batch(5, region="DE")
        WifiShardFactory.create_batch(6, region="US")
        wifi = WifiShardFactory()
        wifi.region = None
        session.add(RegionStat(region="US", blue=1, wifi=2))
        session.add(RegionStat(region="TW", wifi=1))
        session.flush()

        update_statregion.delay().get()
        stats = session.query(RegionStat).all()
        assert len(stats) == 4

        update_statregion.delay().get()
        stats = session.query(RegionStat).all()
        assert len(stats) == 4
```

Reply:
I like independent tests, good idea.

This new test covers that unchanged cell areas don't break ``update_statregion``. This could also be done by setting some ``RegionStat`` rows and calling ``update_statregion``. The existing test covers inserting new ``RegionStat`` rows.

 A few more things to test:

* An updated ``CellArea`` results in an updated ``RegionStat``
* If all the ``CellArea``s for a region are removed, then the ``RegionStat`` is deleted.

I may need to resurrect code coverage to guide these changes.

Reply:
Code coverage is interesting.

For the old code, using ``bulk_update_mappings``, code coverage wouldn't have highlighted any problems. It just shows branch misses for the case with no inserts, updates, and nothing to delete.

For the new code, the existing test covers everything but the case where there are no obsolete regions to delete. Running ``update_statregion`` twice covers this branch. In the original test, the ``TW`` region is an obsolete region that will be deleted.

The existing test includes a change to the ``US`` region. I could add more data that includes a region that does not update, or I could add multiple tests for each case - new region, updated region, same region, deleted region.  I have a preference for single purpose test cases.